### ISSUE TITLE
unable to upload oci-image resource from dockerhub using image digest

### ISSUE THREAD

**Issue by baycarbone** [canonical/charmcraft#444]
Trying to upload an oci-image resource from dockerhub to charmhub for my charm fails if I specify the image digest but is working when giving the image name and tag.

![image](https://user-images.githubusercontent.com/78857322/125141006-19993a80-e114-11eb-834e-285650b54801.png)

Failing with image digest:
```
$ charmcraft upload-resource --image=sha256:ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d  rancher-integrator rancher-integrator-image
The 'ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d' image does not exist in the destination registry (full execution logs in /home/ubuntu/snap/charmcraft/common/charmcraft-log-v3zhisoq)
1 ubuntu@node07:~/charm-rancher-integrator$ cat /home/ubuntu/snap/charmcraft/common/charmcraft-log-v3zhisoq
2021-07-09 22:13:40,472  charmcraft.guard               DEBUG    Starting charmcraft version 1.0.0
2021-07-09 22:13:40,472  charmcraft.main                DEBUG    Raw pre-parsed sysargs: args={'help': False, 'verbose': False, 'quiet': False, 'project_dir': None} filtered=['upload-resource', '--image=sha256:ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d', 'rancher-integrator', 'rancher-integrator-image']
2021-07-09 22:13:40,472  charmcraft.commands            DEBUG    Couldn't find config file /home/ubuntu/charm-rancher-integrator/charmcraft.yaml
2021-07-09 22:13:40,473  charmcraft.main                DEBUG    General parsed sysargs: command='upload-resource' args=['--image=sha256:ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d', 'rancher-integrator', 'rancher-integrator-image']
2021-07-09 22:13:40,474  charmcraft.main                DEBUG    Command parsed sysargs: Namespace(charm_name='rancher-integrator', filepath=None, image=OCIImageSpec(organization='library', name='sha256', reference='ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d'), resource_name='rancher-integrator-image')
2021-07-09 22:13:40,474  charmcraft.commands.store      DEBUG    Uploading resource from image OCIImageSpec(organization='library', name='sha256', reference='ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d') at Dockerhub
2021-07-09 22:13:40,474  charmcraft.commands.store.registry DEBUG    Checking if manifest is already uploaded
2021-07-09 22:13:40,474  charmcraft.commands.store.registry DEBUG    Hitting the registry: HEAD https://registry.hub.docker.com/v2/library/sha256/manifests/ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d
2021-07-09 22:13:40,918  charmcraft.commands.store.registry DEBUG    Authenticating! {'realm': 'https://auth.docker.io/token', 'service': 'registry.docker.io', 'scope': 'repository:library/sha256:pull'}
2021-07-09 22:13:41,736  charmcraft.commands.store.registry DEBUG    Bad response when checking for uploaded 'https://registry.hub.docker.com/v2/library/sha256/manifests/ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d': 401 (headers={'content-type': 'application/json', 'docker-distribution-api-version': 'registry/2.0', 'www-authenticate': 'Bearer realm="https://auth.docker.io/token",service="registry.docker.io",scope="repository:library/sha256:pull",error="insufficient_scope"', 'date': 'Fri, 09 Jul 2021 22:13:41 GMT', 'content-length': '157', 'strict-transport-security': 'max-age=31536000'})
2021-07-09 22:13:41,737  charmcraft                     ERROR    The 'ec7a755e6313da2f7db02d8e82f6b0813b176f06c5622174c8ab45feefc8096d' image does not exist in the destination registry (full execution logs in /home/ubuntu/snap/charmcraft/common/charmcraft-log-v3zhisoq)
```
Working with image name:
```
$ charmcraft upload-resource --image=python:slim-buster  rancher-integrator rancher-integrator-image
Revision 2 created of resource 'rancher-integrator-image' for charm 'rancher-integrator'
```

**Comment by facundobatista** [comment]
Hello @baycarbone, thanks for the report.

One one hand, the important thing here is to understand if you're not blocked. If you was able to upload it using the name and continue with your work, we're fine, but if you're blocked, please let me/us know.

On the other hand, we're about to release `charmcraft 1.1.1` to `stable` (currently in `candidate`), that changes how OCI Images are uploaded: in this new version they will be uploaded from your local machine, always referencing them using the digest. There is more info about it [in the docs](https://juju.is/docs/sdk/publishing-resources).

So, I'm closing this issue, but again, please let me/us know if you're blocked somehow. Regards,

**Comment by baycarbone** [comment]
Thanks for the feedback. I guess this is a case of the documentation being aligned with ```charmcraft 1.1.1``` already as opposed to the current stable version. I followed the doc you linked initially which is why I bumped into this issue.

Depending how soon version ```1.1.1``` will be released, perhaps it might be helpful to add a note in the doc to mention this is for ```1.1.1``` and if you are using the current stable, then use the image name.
